conan/internal/graph/compatibility.py

```diff
@@ -89,17 +89,17 @@ def migrate_compatibility_files(cache_folder):
     compatibility_file = os.path.join(compatible_folder, "compatibility.py")
     cppstd_compat_file = os.path.join(compatible_folder, "cppstd_compat.py")
 
-    def _should_migrate_file(file_path):
+    def _is_migratable(file_path):
         if not os.path.exists(file_path):
             return True
         content = load(file_path)
         first_line = content.lstrip().split("\n", 1)[0]
         return CONAN_GENERATED_COMMENT in first_line
 
-    if _should_migrate_file(compatibility_file) and _should_migrate_file(cppstd_compat_file):
-        if os.path.exists(compatibility_file) and load(compatibility_file) != _default_compat:
+    if _is_migratable(compatibility_file) and _is_migratable(cppstd_compat_file):
+        if not os.path.exists(compatibility_file) or load(compatibility_file) != _default_compat:
```

Comment:
This and indenting the save are the fixed logic - we want to overwrite the compatibility file only when it does not exist, or when it exsists (but it's guarded by the Conan comment, else we would not be here), and its contents are different. See the first added test for that case